File: dcp-solutions/2024-11-05/solution.rb
```ruby
def possible_colors(graph, k)
  n = graph.size
  colors = Array.new(n) { 0 }

  def backtrack(v, used)
    return true if v == n
    color_count = Array.new(k) { 0 }
    (0...k).each do |c|
      next if used.include?(c)

      return false if (0...n).any? { |u| graph[v][u] != 0 && colors[u] == c }

      used << c
      colors[v] = c
      return backtrack(v + 1, used)
    end

    used.each { |c| colors[v] = -1 }
    true
  end

  backtrack(0, [])
end
```

File: dcp-solutions/2024-11-05/solution.ex
```elixir
defmodule GraphColoring do
  @doc """
  Determine whether each vertex in the graph can be colored such that no two adjacent vertices share the same color using at most k colors.

  ## Parameters
  - adjacency_matrix: An undirected graph represented as an adjacency matrix.
  - k: The maximum number of colors to use.

  ## Returns
  A boolean indicating whether a valid coloring is possible.
  """
  @spec graph_coloring(boolean | nil, nonnegative_integer) :: boolean
  def graph_coloring(null, _), do: false

  def graph_coloring(matrix, 0), do: false

  def graph_coloring(matrix, k) when is_binary(matrix), do: graph_coloring(binary_to_integer(matrix), k)

  @spec graph_coloring(integer | binary, nonnegative_integer) :: boolean
  def graph_coloring(matrix, k) when is_integer(matrix) or is_binary(matrix) do
    # Validate input
    if matrix == null || k < 1, do: false

    # Iterate over each vertex in the graph
    for row <- 0..length(matrix) - 1, do: color(row, k)
  end

  defp color(vertex, k) when vertex >= length(matrix), true

  defp color(vert, k) do
    if matrix[vert][vert] == 0, do: true

    # Color the current vertex with a random available color
    available_colors = [i for i <- 1..k]
    new_color = Enum.find(available_colors, fn(i) ->
      valid?(matrix, vert, i)
    end)

    matrix[vert][vert] = new_color

    true
  end

  defp valid?(matrix, vertex, color) do
    for neighbor <- 0..length(matrix) - 1, do: if(matrix[vertex][neighbor] == 1 or (vertex > neighbor and matrix[neighbor][vertex] == 1)), do: false
      else: true
    end
  end
end
```